agenda.php

```diff
@@ -9,15 +9,126 @@ class NotImplementedException extends BadMethodCallException {}
-    public function getEvent(string $url);
-    public function clearEvents();
+abstract class Agenda {
+    protected $caldav_client;
+    protected $api;
+
+    abstract public function __construct(string $vCalendarFilename, string $username, string $password, object $api, array $agenda_args);
+    abstract public function getUserEventsFiltered(string $userid, array $filters_to_apply = array());
+    abstract public function getParsedEvent(string $vCalendarFilename, string $userid);
+    abstract public function getEvents();
+    abstract protected function update(array $ETags);
+    
+    abstract protected function saveEvent(string $vCalendarFilename, string $ETag, object $vCalendar);
+    abstract protected function getEvent(string $vCalendarFilename);
+    abstract public function clearEvents();
+
+    abstract protected function getCTag();
+    abstract protected function setCTag(string $CTag);
+    
+    public function checkAgenda() {
+        $remote_CTag = $this->caldav_client->getCTag();
+        if(is_null($remote_CTag)) {
+            $this->log->error("Fail to update the CTag");
+            return;
+        }
+        
+        // check if we need to update events from the server
+        $local_CTag = $this->getCTag();
+        $this->log->debug("local CTag is $local_CTag, remote CTag is $remote_CTag");
+        if (is_null($local_CTag) || $local_CTag != $remote_CTag){
+            $this->log->debug("Agenda update needed");
+            
+            $remote_ETags = $this->caldav_client->getETags();
+            if($remote_CTag === false || is_null($remote_CTag)) {
+                $this->log->error("Fail to get calendar ETags");
+                return;
+            }
+            
+            $this->update($remote_ETags);
+            $this->setCTag($remote_CTag);
+        }
+
+    }
+    
+    public function updateAttendee(string $vCalendarFilename, string $usermail, bool $add, ?string $attendee_CN) {
+        $this->log->info("updating $vCalendarFilename");
+        list($vCalendar, $ETag) = $this->getEvent($vCalendarFilename);
+        
+        if($add) {
+            if(isset($vCalendar->VEVENT->ATTENDEE)) {
+                foreach($vCalendar->VEVENT->ATTENDEE as $attendee) {
+                    if(str_replace("mailto:","", (string)$attendee) === $usermail) {
+                        if(isset($attendee['PARTSTAT']) && (string)$attendee['PARTSTAT'] === "DECLINED") {
+                            $this->log->info("Try to add a user that have already declined invitation (from outside).");
+                            // clean up
+                            $vCalendar->VEVENT->remove($attendee);
+                            // will add again the user
+                            break;
+                        } else {
+                            $this->log->info("Try to add a already registered attendee");
+                            return 0; // not an error
```

Comment:
Here (and on L99) this method may return something which is not a boolean. I'm not sure it's very intuitive. I can think of some ways to do it differently:

- this method could return `true` if it adds or remove the user from the event, `false` if it didn't need to update the remote event, and throw an exception if an error occured
- or it could return an enum like `{SUCCESS, ERROR, NOOP}`
(arg, when I started writing this I had a 3rd option in mind but now I forgot about it...)

Regardless of what we do with this, I think that the caller of this method does not handle well the case where this one currently returns `0` because it seems that `SlackEvents.register` will try to create or delete a reminder (though it's pointless in that case)

Reply:
What about `{true, false, null}`, instead of `{true, false, 0}`?
Enums will be available in php 8.1 (https://php.watch/versions/8.1/enums, we are running 8.0.13 currently).
Indeed, we need to fix `SlackEvents.register`.
I propose 08fb105.